app/models/user.rb

```diff
@@ -9,9 +9,67 @@ class User < ApplicationRecord
   validates :first_name, :last_name, :username, presence: true
   validates :username, uniqueness: true
 
-  def data
-    User.all.as_json.find do |user|
-      user['username'] == username
+  has_many :active_friendships, foreign_key: :active_friend_id, dependent: :destroy, class_name: 'Friendship'
+  has_many :passive_friendships, foreign_key: :passive_friend_id, dependent: :destroy, class_name: 'Friendship'
+  has_many :active_friends, through: :passive_friendships, source: :active_friend, dependent: :destroy
+  has_many :passive_friends, through: :active_friendships, source: :passive_friend, dependent: :destroy
+
+  default_scope { order(created_at: :asc) }
+
+  def friends
+    User.where(id: active_friendships.where(confirmed: true).pluck(:passive_friend_id))
+      .or(User.where(id: passive_friendships.where(confirmed: true).pluck(:active_friend_id)))
+  end
+
+  def pending_received_requests_from
+    User.where(
+      id: passive_friendships
+        .where(confirmed: false)
+        .pluck(:active_friend_id)
+    )
+  end
+
+  def pending_sent_requests_to
+    User.where(
+      id: active_friendships
+        .where(confirmed: false)
+        .pluck(:passive_friend_id)
+    )
+  end
+
+  def mutual_friends_with(other_user, page)
+    offset = (page.to_i - 1) * 10
+    User
+      .where(id: friends.pluck(:id))
+      .where(id: other_user.friends.pluck(:id))
+      .limit(10)
+      .offset(offset)
+  end
+
+  def user_with_tag(viewed_user)
```

Comment:
Can you read about serializers, e.g. https://itnext.io/a-quickstart-guide-to-using-serializer-with-your-ruby-on-rails-api-d5052dea52c5 (feel free to google more).
It looks like a great place to use them.

Reply:
I have initially used the `active_model_serializers` here but reverted to manually doing the json data responses because I got stuck with how to include the JSON web token in my `user` serialized data. I will try to work on something but how do you suggest I do it? For instance if I have a `user` serializer I can only include attributes which are in the database. Which means I cannot include my JSON web token for user json data unless I make another column in the database for this token (which I don't think is proper to do as I don't wanna make any updates to the user record every time I log in).